### DISCUSSION TITLE
Warning when using Python via pyenv

### DISCUSSION THREAD

**Discussion by cbrnr** [quarto-dev/quarto-cli#8755] [Q&A] [answered]
### Description

I use pyenv to manage my Python environments (including virtual environments). In a Quarto project using a specific pyenv version, everything works fine (meaning that Quarto correctly uses the currently active Python environment as set by pyenv), but I get a warning e.g. when publishing (again, everything works correctly despite of this warning). The same warnings are also popping up in the output of `quarto check`:

```
Quarto 1.4.550
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.550
      Path: /Users/clemens/Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (external install)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/clemens/Library/TinyTeX/bin/universal-darwin
      Version: 2022

[✓] Checking basic markdown render....OK

WARNING: Specified QUARTO_PYTHON '/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short' does not exist.
[✓] Checking Python 3 installation....OK
      Version: 3.12.2
      Path: /Users/clemens/.pyenv/versions/python-intro-short/bin/python3
      Jupyter: 5.7.1
      Kernels: python3

WARNING: Specified QUARTO_PYTHON '/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short' does not exist.
[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.3.2
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library
      knitr: 1.45
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

Is this something that I should be worried about? It is a bit confusing, because everything is detected correctly, so the warning doesn't make sense to me.

**Comment by mcanouil** [comment]
Did you set `QUARTO_PYTHON` somewhere in your project? Perhaps in `_environment`? (https://quarto.org/docs/projects/environment.html)

If you did not set it, it means Quarto find this and since it is valid, go to the next location to find Python.

It should be fine, but better to not have warnings especially regarding configuration such as the Python setup.

**Reply by cbrnr** [reply]
No, I didn't set it explicitly, but the variable is defined in the terminal where Quarto runs and generates the preview, and it points to the correct Python environment:

```
$ env | grep QUARTO_PYTHON
QUARTO_PYTHON=/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short
```

I do have an empty `_quarto.yml` though.

**Reply by mcanouil** [reply]
`QUARTO_PYTHON` is supposed to point to the binary.

Note the path found for the binary during the check:

- `/Users/clemens/.pyenv/versions/python-intro-short/bin/python3`

And the one that seems to be defined in `QUARTO_PYTHON`:

- `/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short`

Even if the above link is correct, it should be:

- `/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short/bin/python3`


Are you sure the two exist? Are they symbolic link or something like that?

**Reply by cbrnr** [reply]
Right, `/Users/clemens/.pyenv/versions/python-intro-short` is a symbolic link to `/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short` (that's the default behavior for venvs created by `pyenv-virtualenv`).

`QUARTO_PYTHON` is not pointing to the Python binary, but to the environment's root directory:

```
❯ ls $QUARTO_PYTHON   
bin  etc  include  lib  share  pyvenv.cfg
```

The binary is located in `$QUARTO_PYTHON/bin`.

I think the main problem is that I never defined `QUARTO_PYTHON`, this seems to be done by the Quarto Visual Studio Code extension. And if it defines it, it should properly point to a Python binary. Yet everything is working fine...

**Reply by mcanouil** [reply]
I don't believe Quarto set this as it's for the user to override whatever auto-detection Quarto does.

I think the issue here is the relative link and Deno.

Could you try to make it as an hard link without touching the environment variable?

**Reply by cbrnr** [reply]
> I don't believe Quarto set this as it's for the user to override whatever auto-detection Quarto does.

But maybe the Quarto extension sets this? I certainly did not set it...

> Could you try to make it as an hard link without touching the environment variable?

I cannot create a hard link to a directory unfortunately.

**Reply by cbrnr** [reply]
I think the problem is really specific to the VS Code extension. In a regular terminal, the output of `quarto check` is fine:

```
[✓] Checking Python 3 installation....OK
      Version: 3.12.2
      Path: /Users/clemens/.pyenv/versions/python-intro-short/bin/python3
      Jupyter: 5.7.1
      Kernels: python3
```

However, when I preview a Quarto document in VS Code (by clicking on the "Preview" icon), a new terminal tab opens, and Quarto runs its preview command (`quarto preview /Users/clemens/Projects/python-intro-short/12/12.qmd --no-browser --no-watch-inputs`). And it's in *this* terminal that it shows the warning!

**Reply by mcanouil** [reply]
Did you try to run quarto within a terminal in VSCode? Maybe that's what you meant by "regular terminal" ...

When I have a bit more time, I'll try to setup a repository to reproduce this, likely using Devcontainer.

**Reply by cbrnr** [reply]
Yes, by regular terminal I mean any terminal running my default shell, including terminals opened in VS Code. It's just that Quarto Preview opens a special terminal, where Python detection produces this weird warning.

**Reply by cderv** [reply, marked as answer]
Let me clarifiy some things here: 

> QUARTO_PYTHON is supposed to point to the binary.

@mcanouil we support `QUARTO_PYTHON` has a directory in which the binary can be found. If a directory is provided we look for these binaries (`const bin = ["python3", "python", "python3.exe", "python.exe"]`). You can check the resolution of `QUARTO_PYTHON` in the code base. 

> But maybe the Quarto extension sets this? I certainly did not set it...

@cbrnr Yes in VSCODE we do set the `QUARTO_PYTHON` env var to use the active Python Interpreter. This can be found in the code base for the extension
https://github.com/quarto-dev/quarto/blob/f9edec34e53817e9c9f73cb349f524a82bcc6474/apps/vscode/src/core/terminal.ts#L40-L44

What is your active python interpreter in VSCODE ? Is this `/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short` (probably) ? 

Is this a working version in VSCODE for the python extension ? Are there binaries in `/Users/clemens/.pyenv/versions/3.12.2/envs/python-intro-short` ? 

When `QUARTO_PYTHON` value is not found, it just throws a warning (that you see) and goes on with the usual detection looking for python in PATH. I believe this is where `/Users/clemens/.pyenv/versions/python-intro-short/bin/python3` is found. 

You should check which Python version is configured to be used in VSCODE using the python extension. This python version is meant to be used with python cells or any Python work you would do in VSCODE 

Here is a doc about this Python interpreter thing: https://code.visualstudio.com/docs/python/environments

Hope it helps

**Reply by cbrnr** [reply]
Thanks @mcanouil and @cderv for your help! I think I could track it down to the currently selected Python interpreter in VS Code not being set correctly. I re-selected the same venv, and now everything works fine! Thanks a lot! 🙏

**Reply by mcanouil** [reply]
Glad you fixed your issue.
By the way, why are you setting the path in VSCode exactly?
I did not and this select the proper Python to run every single time, from Virtual Env, Conda, global.
<img width="459" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/f2ba1d25-1cd8-4a5b-9218-297dfa9d3632">

**Reply by cbrnr** [reply]
You mean you never select the project interpreter? It prompts me every time I create a new project. This setting was set to some venv I created previously, and I'm not sure VS Code automatically picks up the correct Python env I set with a `.python-version` file.

**Reply by cbrnr** [reply]
I just tried setting it to `python` like you have, and it does seem to recognize the correct env! Thanks! 💯

**Reply by mcanouil** [reply]
It's the default value by the way.

I get prompted to select when creating environment for example and in some cases with Jupyter notebooks.
But Python REPL always take my activated environment (same for Quarto, etc.).

**Reply by cbrnr** [reply]
Yeah, somehow this setting was changed to some weird value, and I never noticed (and just manually selected the right env). I hope it now stays like this 😄!
